src/Engine/System/RenderSystem.cpp

```diff
@@ -20,24 +26,57 @@ void RenderSystem::update(float dt) {
     glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
     glClearColor(0.2f, 0.3f, 0.4f, 1.f);
 
-    /* Loop through active shaders */
-    for (auto &shader : m_shaders) {
-        if (!shader.second->isEnabled()) {
-            continue;
-        }
+    if (s_camera) {
+        /* Loop through active shaders */
+        for (auto &shader : s_shaders) {
+            if (!shader.second->isEnabled()) {
+                continue;
+            }
 
-        shader.second->bind();
-        ///////////////////////////  TODO  ///////////////////////////
-        // pass a list of render components that are specific       //
-        // to this shader -- right now we are passing the entire    //
-        // list and expecting each shader to filter through         //
-        //////////////////////////////////////////////////////////////
-        shader.second->render(m_components);
-        shader.second->unbind();
+            shader.second->bind();
+            ///////////////////////////  TODO  ///////////////////////////
+            // pass a list of render components that are specific       //
+            // to this shader -- right now we are passing the entire    //
+            // list and expecting each shader to filter through         //
+            //////////////////////////////////////////////////////////////
+
+            // this reinterpret_cast business works because unique_ptr's data is
+            // guaranteed is the same as a pointer
+            shader.second->render(*s_camera, reinterpret_cast<std::vector<Component *> &>(s_diffuseComponents));
+            shader.second->unbind();
+        }
     }
 
     /* ImGui */
     if (Window::isImGuiEnabled()) {
         ImGui::Render();
     }
+}
+
+void RenderSystem::add(Component & component) {
+    if (dynamic_cast<DiffuseRenderComponent *>(&component))
```

Comment:
These dynamic casts in add(), remove() and update() in systems need to happen for ever new component that is introduced? Some documentation about this would be nice.

Reply:
Here the system is being passed a generic Component and is expected to store it as a derived type, in this case DiffuseRenderComponent. The dynamic casts are there to figure out what its derived type it is, or if it's not a type belonging to that system, to error. I'm actually changing this back to having Scene store components, as this makes removing components faster, so this will go away

Reply:
I'll merge after these changes